integrations/nifi/nifi-rest/nifi-rest-client/nifi-rest-client-api/src/main/java/com/thinkbiganalytics/nifi/rest/client/AbstractNiFiControllerServicesRestClient.java

```diff
@@ -81,6 +95,161 @@ public ControllerServiceDTO updateStateById(@Nonnull final String id, @Nonnull f
+
+        //Stop all processor references and also disable any other controller service references
+        if (references.isPresent()) {
+            //build the reference state and revision maps prior to making this update
+            referencingComponentEntityMap = getReferencingComponents(references.get().getControllerServiceReferencingComponents());
+
+            previousProcessorState =
+                referencingComponentEntityMap.values().stream().filter(e -> e.getComponent().getReferenceType().equalsIgnoreCase("PROCESSOR")).map(c -> c.getComponent()).collect(Collectors.toMap(
+                    ControllerServiceReferencingComponentDTO::getId, ControllerServiceReferencingComponentDTO::getState));
+
+            processorRevisions = referencingComponentEntityMap.values().stream().filter(e -> e.getComponent().getReferenceType().equalsIgnoreCase("PROCESSOR")).collect(Collectors.toMap(
+                ControllerServiceReferencingComponentEntity::getId, ControllerServiceReferencingComponentEntity::getRevision));
+
+            controllerServiceRevisions =
+                referencingComponentEntityMap.values().stream().filter(e -> e.getComponent().getReferenceType().equalsIgnoreCase("ControllerService")).collect(Collectors.toMap(
+                    ControllerServiceReferencingComponentEntity::getId, ControllerServiceReferencingComponentEntity::getRevision));
+
+            //Stop the referencing processors and ensure they are in the stopped state
+            if (!processorRevisions.isEmpty()) {
+                String state = NifiProcessUtil.PROCESS_STATE.STOPPED.name();
+                log.info("Stopping all component references to controller service {} ", controllerService.getName());
+                UpdateControllerServiceReferenceRequestEntity stopComponentsRequest = new UpdateControllerServiceReferenceRequestEntity();
+                stopComponentsRequest.setState(state);
+                stopComponentsRequest.setId(controllerService.getId());
+                stopComponentsRequest.setReferencingComponentRevisions(processorRevisions);
+                updateReferences(id, stopComponentsRequest);
+                boolean updatedReferences = ensureComponentsAreOfState(id, "Processor", state, 5, 500, TimeUnit.MILLISECONDS);
+                if (!updatedReferences) {
+                    //error unable to change the state of the references. ... error
+                    throw new NifiClientRuntimeException("Unable to stop processor references to this controller service " + controllerService.getName() + " before making the update");
+                }
+            }
+
+            //disable any controller service references
+            if (!controllerServiceRevisions.isEmpty()) {
+                UpdateControllerServiceReferenceRequestEntity stopComponentsRequest = new UpdateControllerServiceReferenceRequestEntity();
+                stopComponentsRequest.setState("DISABLED");
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "DISABLED" 4 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)